### PR TITLE
Track contract elements 

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
This PR adds a `contract_elements` table which contains elements for all contracts in our database and updates their accumulator proofs. This a requirement for broadcasting file contract related txns such as broadcasting revisions.

### CODE REVIEW

File: internal/bus/chainsubscriber.go [hunk 1]
```diff
@@ -221,15 +237,30 @@ func (s *chainSubscriber) revertChainUpdate(tx sql.ChainUpdateTx, cru chain.Reve
...
+	cru.ForEachV2FileContractElement(func(fce types.V2FileContractElement, created bool, rev *types.V2FileContractElement, res types.V2FileContractResolutionType) {
+		if created && s.isKnownContract(fce.ID) {
+			revertedContracts = append(revertedContracts, fce.ID)
+		}
 		cus = append(cus, v2ContractUpdate(fce, rev, res))
 	})
+
+	// reverted contracts
+	if err := tx.RemoveFileContractElements(revertedContracts); err != nil {
+		return fmt.Errorf("failed to insert v2 file contract elements: %w", err)
```
Comment (peterjan) [inline internal/bus/chainsubscriber.go:250]:
```suggestion
		return fmt.Errorf("failed to remove v2 file contract elements: %w", err)
```

File: internal/sql/migrations.go [hunk 2]
```diff
@@ -373,6 +373,12 @@ var (
 					return performMigration(ctx, tx, migrationsFs, dbIdentifier, "00028_contract_usability", log)
 				},
 			},
+			{
+				ID: "00029_contract_elements",
```
Comment (peterjan) [inline internal/sql/migrations.go:377]: I'll race you for it

File: stores/chain_test.go [hunk 3]
```diff
@@ -391,3 +393,74 @@ func TestProcessChainUpdate(t *testing.T) {
...
+				},
+				V2FileContract: contract,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		assertContractElement(tx, 1, []types.Hash256{{1}})
+
+		// TODO: update the element's proof
```
Comment (peterjan) [inline stores/chain_test.go:453]: leftover? we can already implement this test with what's currently available to us right?

File: stores/sql/chain.go [hunk 4]
```diff
@@ -118,26 +119,141 @@ func UpdateWalletSiacoinElementProofs(ctx context.Context, tx sql.Tx, updater wa
...
+		}
+	}
+	return nil
+}
+
+func FileContractElement(ctx context.Context, tx sql.Tx, fcid types.FileContractID) (types.V2FileContractElement, error) {
+	var contract V2Contract
+	var leafIndex uint64
+	var proof MerkleProof
+	err := tx.QueryRow(ctx, "SELECT contract, leaf_index, merkle_proof FROM contract_elements ce INNER JOIN contracts c ON ce.db_contract_id = c.id WHERE c.fcid = ?", Hash256(fcid)).
```
Comment (peterjan) [inline stores/sql/chain.go:163]: same thing but better use `FileContractID` here

File: stores/sql/chain.go [hunk 5]
```diff
@@ -118,26 +119,141 @@ func UpdateWalletSiacoinElementProofs(ctx context.Context, tx sql.Tx, updater wa
...
+
+	for _, fce := range fces {
+		var contractID int64
+		err = contractIDStmt.QueryRow(ctx, Hash256(fce.ID)).Scan(&contractID)
+		if errors.Is(err, dsql.ErrNoRows) {
+			return contractNotFoundErr(fce.ID)
+		} else if err != nil {
+			return err
+		}
+		_, err = insertStmt.Exec(ctx, time.Now().UTC(), contractID, V2Contract(fce.V2FileContract), fce.StateElement.LeafIndex, MerkleProof{fce.StateElement.MerkleProof})
```
Comment (peterjan) [inline stores/sql/chain.go:199]: `nit` use `time.Now()` to be consistent

File: stores/sql/chain.go [hunk 6]
```diff
@@ -118,26 +119,141 @@ func UpdateWalletSiacoinElementProofs(ctx context.Context, tx sql.Tx, updater wa
...
+
+	deleteStmt, err := tx.Prepare(ctx, "DELETE FROM contract_elements WHERE db_contract_id = ?")
+	if err != nil {
+		return err
+	}
+	defer deleteStmt.Close()
+
+	for _, fcid := range fcids {
+		var contractID int64
+		if err := contractIDStmt.QueryRow(ctx, Hash256(fcid)).Scan(&contractID); errors.Is(err, dsql.ErrNoRows) {
```
Comment (peterjan) [inline stores/sql/chain.go:222]: use `FileContractID`

File: stores/sql/chain.go [hunk 7]
```diff
@@ -118,26 +119,141 @@ func UpdateWalletSiacoinElementProofs(ctx context.Context, tx sql.Tx, updater wa
...
+	}
+	defer contractIDStmt.Close()
+
+	deleteStmt, err := tx.Prepare(ctx, "DELETE FROM contract_elements WHERE db_contract_id = ?")
+	if err != nil {
+		return err
+	}
+	defer deleteStmt.Close()
+
+	for _, fcid := range fcids {
```
Comment (peterjan) [inline stores/sql/chain.go:220]: is this really necessary? I think you can argue it's safe to do `DELETE FROM WHERE id IN (SELECT id FROM ` here..

**Review by peterjan — COMMENTED** [review]
left some remarks but overall looks good to go 👍

File: stores/sql/chain.go (hunk 7 above)
Comment (ChrisSchinnerl) [reply stores/sql/chain.go:220]: Theoretically we could probably change it but that means building the placeholder string which isn't nice either. This feels a bit more readable and obvious. We also don't need to worry about performance here.

**Review by peterjan — APPROVED** [review]

**Pull request merged**